src/dataflow/src/render/reduce.rs

```diff
@@ -396,6 +448,140 @@ where
+    partial.reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceMinsMaxes", {
+        let mut row_packer = RowPacker::new();
+        move |key, source, target| {
+            // Negative counts would be surprising, but until we are 100% certain we wont
+            // see them, we should report when we do. We may want to bake even more info
+            // in here in the future.
+            if source.iter().any(|(_val, cnt)| cnt < &0) {
+                // XXX: This reports user data, which we perhaps should not do!
+                for (val, cnt) in source.iter() {
+                    if cnt < &0 {
+                        log::error!("[customer-data] Negative accumulation in ReduceMinsMaxes: {:?} with count {:?}", val, cnt);
+                    }
+                }
+            } else {
+                // Pack the value with the key as the result.
+                if prepend_key {
+                    row_packer.extend(key.iter());
+                }
+                for (aggr_index, func) in aggr_funcs.iter().enumerate() {
+                    let iter = source.iter().map(|(values, _cnt)| values[aggr_index].iter().next().unwrap());
+                    row_packer.push(func.eval(iter, &RowArena::new()));
+                }
+                target.push((row_packer.finish_and_reuse(), 1));
+            }
+        }
+    })
+}
+
+fn build_mins_maxes_stage<G>(
+    collection: Collection<G, ((Row, u64), Vec<Row>)>,
+    aggrs: Vec<AggregateFunc>,
+    modulus: u64,
+) -> Collection<G, ((Row, u64), Vec<Row>)>
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+{
+    let input = collection.map(move |((key, hash), values)| ((key, hash % modulus), values));
+
+    let negated_output = input
+        .reduce_named("MinsMaxesHierarchical", {
+            let mut row_packer = repr::RowPacker::new();
+            move |key, source, target| {
+                // Should negative accumulations reach us, we should loudly complain.
+                if source.iter().any(|(_val, cnt)| cnt <= &0) {
+                    for (val, cnt) in source.iter() {
+                        if cnt <= &0 {
+                            // XXX: This reports user data, which we perhaps should not do!
+                            log::error!("[customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: {:?}\tvalue: {:?}\tcount: {:?}", key, val, cnt);
+                        }
+                    }
+                } else {
+                    // We ignore the count here under the belief that it cannot affect
+                    // hierarchical aggregations; should that belief be incorrect, we
+                    // should certainly revise this implementation.
+
+                    let mut output = Vec::with_capacity(aggrs.len());
+                    for (aggr_index, func) in aggrs.iter().enumerate() {
+                        let iter = source.iter().map(|(values, _cnt)| values[aggr_index].iter().next().unwrap());
```

Comment:
values is a vec of rows instead of a single row in order to have random access and make this reduction the same amount of work as doing unfused reductions separately. not sure if its actually worth storing a vec of rows / having to have a new function deal with it